### PR TITLE
Relax error when using SWIFT_RETURNS_(UN)RETAINED on templated return types resolving to non-frts

### PR DESCRIPTION
**Pull request by fahadnayyar**
We should not complain about usage of `SWIFT_RETURNS_(UN)RETAINED` for templated C++ APIs when instantiated with a non `SWIFT_SHARED_REFERENCE` types

rdar://143732201

### CODE REVIEW

**Comment by fahadnayyar** [comment]
@swift-ci please smoke test

File: lib/ClangImporter/ImportDecl.cpp [hunk 1]
```diff
@@ -3427,6 +3427,12 @@ namespace {
         }
       } else {
         if (returnsRetainedAttrIsPresent || returnsUnretainedAttrIsPresent) {
+          if (const auto *functiionDecl = dyn_cast<clang::FunctionDecl>(decl)) {
```
Comment (Xazax-hun) [inline lib/ClangImporter/ImportDecl.cpp:3430]: Nit: spelling mistake in the variable name.

File: lib/ClangImporter/ImportDecl.cpp [hunk 2]
```diff
@@ -3427,6 +3427,12 @@ namespace {
...
+            if (functiionDecl->getTemplateSpecializationKind() !=
+                clang::TSK_Undeclared) {
```
Comment (Xazax-hun) [inline lib/ClangImporter/ImportDecl.cpp:3432]: I am not actually sure if this does what you mean. According to the documentation `TSK_Undeclared` means:
"This template specialization was formed from a template-id but has not yet been declared, defined, or instantiated."

We have a `isTemplateInstantiation` method. Would that work?

**Review by Xazax-hun — CHANGES_REQUESTED** [review]

File: lib/ClangImporter/ImportDecl.cpp (hunk 2 above)
Comment (fahadnayyar) [reply lib/ClangImporter/ImportDecl.cpp:3432]: Yes, `isTemplateInstantiation` should work fine!

**Comment by fahadnayyar** [comment]
@swift-ci please smoke test

File: test/Interop/Cxx/foreign-reference/frt-retained-unretained-attributes-error.swift [hunk 3]
```diff
@@ -53,3 +53,7 @@ let w = x - y
 let f = FunctionVoidToFRTStruct()
 let frt = f()
 // CHECK-NOT: warning: 'operator()' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it is returning a SWIFT_SHARED_REFERENCE
+
+let nonFrt = NonFRTStruct()
+let nonFrtLocalVar1 = global_function_returning_templated_retrun_frt_owned(nonFrt)
+// CHECK-NOT: error: 'global_function_returning_templated_retrun_frt_owned<NonFRTStruct>' cannot be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED because it is not returning a SWIFT_SHARED_REFERENCE type
```
[the remark this answers is not in the document]
Comment (Xazax-hun) [reply test/Interop/Cxx/foreign-reference/frt-retained-unretained-attributes-error.swift:53]: I still see filecheck directives in this file as opposed to `-verify` expectations so unresolving this thread for now.

[the remark this answers is not in the document]
Comment (fahadnayyar) [reply test/Interop/Cxx/foreign-reference/frt-retained-unretained-attributes-error.swift:53]: Sorry I forgot to delete those directives.

**Comment by fahadnayyar** [comment]
@swift-ci please smoke test

File: test/Interop/Cxx/foreign-reference/Inputs/cxx-functions-and-methods-returning-frt.h [hunk 4]
```diff
@@ -4,9 +4,9 @@
 // FRT or SWIFT_SHARED_REFERENCE type
 struct FRTStruct {
   // Friend function declarations
-  friend FRTStruct *returnInstanceOfFRTStruct(int v);
-  friend FRTStruct *returnInstanceOfFRTStructWithAttrReturnsRetained(int v);
-  friend FRTStruct *returnInstanceOfFRTStructWithAttrReturnsUnretained(int v);
+  friend FRTStruct *returnInstanceOfFRTStruct(int v); // expected-warning {{'returnInstanceOfFRTStruct' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it is returning a SWIFT_SHARED_REFERENCE}}
```
Comment (egorzhdan) [inline test/Interop/Cxx/foreign-reference/Inputs/cxx-functions-and-methods-returning-frt.h:7]: Not sure I follow – the commit message mentions relaxing the diagnostic, but it looks like there are extra diagnostics emitted in this patch?

Comment (fahadnayyar) [reply test/Interop/Cxx/foreign-reference/Inputs/cxx-functions-and-methods-returning-frt.h:7]: @egorzhdan these diagnostics are coming here because of lit test refactoring. We changed the to use  `-verify`  and `-verify-additional-file` instead of `%FileCheck %s`.

This patch does suppress the error when using `SWIFT_RETURNS_(UN)RETAINED` annotation in case of templated return for cases where return type eventually resolved to a C++ foreign reference type.

Like for this case that I added in the lit tests:

// C++
```
struct NonFRTStruct {};

template <typename T>
T global_function_returning_templated_retrun_frt_owned(T)
    __attribute__((swift_attr("returns_retained")));
```


// Swift
```
let nonFrt = NonFRTStruct()
let nonFrtLocalVar1 = global_function_returning_templated_retrun_frt_owned(nonFrt)
```

Why I missed this edge case in original implementation?? Lack of Testing!!

In the original patch I tested that the semantics (behaviour) of `global_function_returning_templated_retrun_frt_owned` is correct/as expected when calling with a C++ frt v/s value type:

// Swift
```
    let frtLocalVar14 = global_function_returning_templated_retrun_frt_owned(frt)
    // CHECK: function_ref @{{.*}}global_function_returning_templated_retrun_frt_owned{{.*}} : $@convention(c) (FRTStruct) -> @owned FRTStruct

    let nonFrtLocalVar1 = global_function_returning_templated_retrun_frt_owned(nonFrt)
    // CHECK: function_ref @{{.*}}global_function_returning_templated_retrun_frt_owned{{.*}} : $@convention(c) (NonFRTStruct) -> NonFRTStruct
```

This is another example demonstrating the importing of "shifting left" in software engineering. [Reference](https://learn.microsoft.com/en-us/devops/develop/shift-left-make-testing-fast-reliable)

We should invest in better unit tests, integration tests and production tests (adopters using our compiler features and bug fixes more rapidly) for Swift/C++ Interoperability. Also maybe we should focus on suggesting more test cases when reviewing patches which could help to explore various edge cases to find such bugs earlier (**shifting left philosophy**: unit tests >> integration test >> production test >>>>>> bug shipped to customers).

CC: @Xazax-hun @j-hui @susmonteiro @ravikandhadai

**Comment by fahadnayyar** [comment]
@swift-ci please smoke test

**Comment by fahadnayyar** [comment]
@swift-ci please smoke test

**Review by Xazax-hun — APPROVED** [review]
The relaxation itself looks good to me but I think we are missing some important diagnostics here.

Specifically, I wonder if we should have a warning when a templated function is instantiated with a foreign reference type in a way that the instantiation returns the FRT, but the original templated function is not annotated. Can be a follow-up PR though.

**Review by egorzhdan — APPROVED** [review]

**Comment by fahadnayyar** [comment]
Enabling auto merge with the intention of fixing up the Nits while implementing Gabor's amazing suggestion about the diagnostic in a follow up patch: rdar://148829496
